File: fileviewer/kernel/EventBus.js
export class EventBus {
  constructor() {
    this._handlers = new Map()  // event -> [{handler, pluginId}]
  }

  on(event, handler, pluginId) {
    if (!this._handlers.has(event)) this._handlers.set(event, [])
    this._handlers.get(event).push({ handler, pluginId })
  }

  off(event, handler) {
    const list = this._handlers.get(event)
    if (!list) return
    const idx = list.findIndex(h => h.handler === handler)
    if (idx >= 0) list.splice(idx, 1)
  }

  emit(event, payload) {
    const list = this._handlers.get(event)
    if (!list) return
    for (const { handler } of [...list]) {
      try { handler(payload) } catch (e) { console.error(`EventBus handler error [${event}]:`, e) }
    }
  }

  offPlugin(pluginId) {
    for (const [event, list] of this._handlers) {
      this._handlers.set(event, list.filter(h => h.pluginId !== pluginId))
    }
  }
}

File: fileviewer/kernel/event_bus.py
import asyncio
from collections import defaultdict
from typing import Callable, Any


class EventBus:
    def __init__(self):
        self._async_handlers: dict[str, list[tuple[Callable, str]]] = defaultdict(list)
        self._sync_handlers:  dict[str, list[tuple[Callable, str]]] = defaultdict(list)

    def subscribe(self, event: str, handler: Callable, plugin_id: str) -> None:
        self._async_handlers[event].append((handler, plugin_id))

    def subscribe_sync(self, event: str, handler: Callable, plugin_id: str) -> None:
        self._sync_handlers[event].append((handler, plugin_id))

    async def publish(self, event: str, payload: Any = None) -> None:
        for handler, _ in list(self._sync_handlers.get(event, [])):
            try:
                handler(payload)
            except Exception:
                pass
        coros = []
        for handler, _ in list(self._async_handlers.get(event, [])):
            result = handler(payload)
            if asyncio.iscoroutine(result):
                coros.append(result)
        if coros:
            await asyncio.gather(*coros, return_exceptions=True)

    def unsubscribe_plugin(self, plugin_id: str) -> None:
        for d in (self._async_handlers, self._sync_handlers):
            for event in list(d.keys()):
                d[event] = [(h, pid) for h, pid in d[event] if pid != plugin_id]
